{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cf5bc135-7355-4b67-8ad5-6ffb96f4306e",
   "metadata": {},
   "source": [
    "# Exercises Week 9: Temperature regulated gene expression\n",
    "\n",
    "**Course**: [Topics in life sciences engineering](https://moodle.epfl.ch/enrol/index.php?id=17061) (BIO-411)\n",
    "\n",
    "**Professors**:  _Gönczy Pierre_, _Naef Felix_, _McCabe Brian Donal_\n",
    "\n",
    "SSV, MA, 2024-2025\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b79f4db",
   "metadata": {},
   "source": [
    "This week's exercises are inspired by the work of [Gotic et al., 2015](http://genesdev.cshlp.org/content/30/17/2005), in which RNA levels, including pre-mRNA and mRNA (total RNA-seq), were measured in mouse NIH-3T3 fibroblasts after cells were switched from a warm to a colder temperature (38C° to 33C°), and vice versa (33C° to 38C°).\n",
    "The experiments were conducted over multiple time points in an \"*approach to steady-state*\" design. The cold-inducible *Cirbp* gene was used as an example to examine the kinetics and mechanisms of post-transcritional control of gene expression. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f854f74e",
   "metadata": {},
   "source": [
    "## Setup the Jupyter environment \n",
    "In order to run this notebook, we advise you to use the EPFL's JupyterLab centralized platform [noto.epfl.ch](http://noto.epfl.ch). Some of the required libraries are not available on the platform but you can install them copy-pasting the following commands in the **Jupyter Terminal** (open a new tab with the '+' icon and select \"Terminal\"):\n",
    "\n",
    "````\n",
    "my_venvs_create Week9_env\n",
    "my_venvs_activate Week9_env\n",
    "pip install gseapy\n",
    "pip install adjustText\n",
    "my_kernels_create Week9_env “Week9”\n",
    "my_venvs_deactivate\n",
    "````\n",
    "Refresh your browser's page and select $\\textbf{\\color{red}the \"Week9\" kernel}$ using the top right toggle button. Import the libraries running the code cell below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "073fb31d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Import libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import scale\n",
    "from scipy.integrate import odeint\n",
    "\n",
    "import plotly.express as px\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import gseapy as gp\n",
    "from ipywidgets import interact, fixed\n",
    "from adjustText import adjust_text\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3430d1da-c36f-4886-8a9f-bf1a570c299c",
   "metadata": {},
   "source": [
    "## **Temperature regulates splicing efficiency of the cold-inducible RNA-binding protein gene *Cirbp***\n",
    "This week’s exercises are based on the work of [Gotic et al., 2015](http://genesdev.cshlp.org/content/30/17/2005), which explores how core body temperature (CBT) rhythms influence diurnal gene expression, focusing on the role of the protein Cirbp.\n",
    "\n",
    "In mammals, core body temperature (CBT) fluctuates diurnally around a mean value of 36°C–37°C. While the period of circadian rhythms is stable under temperature changes, the phase of the cycle is highly sensitive to (even small) temperature shifts. For instance, in cultured fibroblasts, the phase of circadian gene expression can be synchronized with simulated CBT cycles.\n",
    "\n",
    "The paper investigates how CBT rhythms might synchronize gene expression, possibly through temperature-sensitive proteins. One such protein, Cirbp (Cold-Inducible RNA-Binding Protein), shows increased mRNA levels at lower temperatures. The authors aim to understand the mechanisms behind this temperature-sensitive response.\n",
    "\n",
    "They explore different ways mRNA regulation could change with temperature, including:\n",
    "- Increased Cirbp transcription at lower temperatures\n",
    "- Reduced degradation of Cirbp mRNA at lower temperatures\n",
    "- Enhanced pre-mRNA splicing efficiency at lower temperatures\n",
    "  \n",
    "Through various experiments (RNase protection assays, ChIP assays, luminescence assays), the authors rule out transcriptional and degradation regulation, showing that increased pre-mRNA splicing efficiency is key to the higher Cirbp mRNA accumulation at lower temperatures.\n",
    "\n",
    "**Experimental design**\n",
    "In these exercises, we will focus on one of the experiments that led to this conclusion. The authors used an “approach to steady-state” (ATSS) RNA-seq technique to examine genome-wide transcription and mRNA accumulation after temperature changes.  Specifically, NIH3T3 cells were exposed to a temperature change (either from 33°C to 38°C or from 38°C to 33°C) and collected at various time points after the shift. Total RNA was then extracted, ribosomal RNA was removed, and cDNA libraries were sequenced. This method enabled the authors to measure both pre-mRNA and mRNA levels across the entire genome.\n",
    "\n",
    "<img src=\"./Experiment.png\" alt=\"Alt text\" width=\"50%\"/>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6949409b",
   "metadata": {},
   "source": [
    "## Exercise 1: Simulation of a simplified production-decay ODE model of transcription"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f03f2bf-5f08-4957-8535-3eb7162a0a47",
   "metadata": {
    "tags": []
   },
   "source": [
    "We first consider a system in which nuclear pre-mRNA is transcribed and subsequently spliced to produce mRNA.  \n",
    "Both molecular species are subject to distinct degradation processes, which can be modeled with a two-dimensional system of ordinary differential equations (ODEs):\n",
    "\n",
    "\\begin{aligned}\n",
    "&\\frac{dP}{dt} = s - (k_p + \\rho)P \\\\\n",
    "&\\frac{dM}{dt} = \\rho P - k_m M \\\\\n",
    "\\end{aligned}\n",
    "\n",
    "where $P$ and $M$ represent the concentrations of pre-mRNA and mRNA, respectively.\n",
    "\n",
    "Here, $s$ the transcription rate, is assumed to be a constant, but could also be a time-dependent function $s(t)$ (this will be explored further in the exercise session for Week 11).  \n",
    "The parameters $k_m$ and $k_p$ are the degration rates of $M$ and $P$, respectively, while $\\rho$ is the splicing rate of pre-mRNA $P$ into mRNA $M$; these rates are also constants.  \n",
    "When all parameters are constant, the system will reach steady-state levels after a transient period, meaning that the derivatives $\\frac{dP}{dt}$ and $\\frac{dM}{dt}$ will equal to zero."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04963fd2-d695-42ca-b65c-4a416a71585e",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Question 1\n",
    "1. Derive the steady-state concentrations and interpret what happens to $M$ when $\\rho\\gg k_p$.  \n",
    "\n",
    "2. Compute the steady-state ratio mRNA/pre-mRNA.   \n",
    "    a. Comment on the result, in particular discuss which parameters are important and which are not.  \n",
    "    b. Discuss how this ratio might be used to compare mRNA half-lives across different conditions.\n",
    "\n",
    "3. Using $\\rho = 2 h^{-1}$, $s = 1 [\\text{P}]h^{-1}$ and $k_m = 0.1 h^{-1}$ (values from litterature), estimate the ratio of mRNA to pre-mRNA molecules in typical mammalian cells."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48cf6713",
   "metadata": {},
   "source": [
    "### Answer 1\n",
    "\n",
    "1. *Type your answers here*  \n",
    "\n",
    "2. *Type your answers here*  \n",
    "\n",
    "3. *Type your answers here*\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8722045",
   "metadata": {},
   "source": [
    "### Question 2 \n",
    "Simulate the production-decay ODE with a constant transcription rate $s$ using the provided code:\n",
    "1. Using the interactive widget (read the code below), describe the effects obtained by varying parameters such as splicing rate ($\\rho$), degradation rates ($k_p$ and $k_m$) and transcription rate ($s$).\\\n",
    "Experiment with different initial conditions for the pre-mRNA and mRNA concentrations as well.\n",
    "\n",
    "2. Adapt the code by adding a few lines to plot the steady-state concentrations of pre-mRNA and mRNA, as derived in Question 1.1. \n",
    "\n",
    "3. Optional: Extend the model by incorporating a third differential equation that represents the protein level (G). Update the simulation and the corresponding plots to include protein dynamics. How similar/different is the relation between P and M compared to the one between M and G ?\n",
    "\n",
    "Remember to clearly label all axes and legends on your plots for easy interpretation."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38c6de31",
   "metadata": {},
   "source": [
    "### Answer 2  \n",
    "\n",
    "1. *Type your answer here*  \n",
    "\n",
    "2. *Type your answer here*  \n",
    "\n",
    "3. *Type your answer here*  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e725faf",
   "metadata": {},
   "source": [
    "#### Definition of important functions\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "deefe0e2-66c6-4ba1-a84f-772cd6357612",
   "metadata": {},
   "outputs": [],
   "source": [
    "def production_decay_ode(y, time, s, kp, km, rho):\n",
    "\n",
    "    \"\"\"\n",
    "    Defines the ODEs for the production and decay of pre-mRNA and mRNA.\n",
    "    P: pre-mRNA concentration\n",
    "    M: mRNA concentration\n",
    "    s: transcription rate (constant)\n",
    "    kp: pre-mRNA degradation rate\n",
    "    km: mRNA degradation rate\n",
    "    rho: splicing rate\n",
    "    \"\"\"\n",
    "    \n",
    "    P, M = y\n",
    "    dPdt = s - (kp + rho) * P\n",
    "    dMdt = rho * P - km * M\n",
    "\n",
    "    return [dPdt, dMdt] \n",
    "\n",
    "\n",
    "def simulation(ode_function, parameters, initial_conditions = None, time = None):\n",
    "\n",
    "    \"\"\"\n",
    "    Simulate the ODEs for the production and decay of pre-mRNA and mRNA.\n",
    "    parameters: parameters for the ODEs\n",
    "    initial_conditions: initial conditions for the ODEs\n",
    "    time: time intervals for the simulation\n",
    "    \"\"\"\n",
    "\n",
    "    s = parameters['s']\n",
    "    kp = parameters['kp']\n",
    "    km = parameters['km']\n",
    "    rho = parameters['rho']\n",
    "\n",
    "    # Placeholder for adding steady-state solutions\n",
    "    # P_ss = ... #Add your steady-state solution for P here \n",
    "    # M_ss = ... #Add you steady-state solution for M here\n",
    "    \n",
    "    results = odeint(ode_function, initial_conditions, time, args=(s, kp, km, rho))\n",
    "\n",
    "    plt.figure(figsize=(10, 6))\n",
    "    plt.ylim([0,150])\n",
    "\n",
    "    # Uncomment the lines below after adding your steady-state solutions\n",
    "    #plt.axhline(y=P_ss, color='blue', linestyle='--', label='P_ss')  \n",
    "    #plt.axhline(y=M_ss, color='orange', linestyle='--', label='M_ss')\n",
    "\n",
    "    plt.title('Simulation of pre-mRNA and mRNA Production-Decay Dynamics')\n",
    "    plt.plot(time, results[:, 0], label = 'P(t)', linewidth=2)\n",
    "    plt.plot(time, results[:, 1], label = 'M(t)', linewidth=2)\n",
    "    plt.xlabel(\"Time [min]\")\n",
    "    plt.ylabel(\"Concentration\")\n",
    "    plt.legend()\n",
    "    plt.show()\n",
    "    \n",
    "def animate(initial_conditions, time, s=5, kp= 0.05, km=0.05, rho=0.4):\n",
    "    \"\"\"\n",
    "    Runs the simulation with an interactive widgets for parameter exploration.\n",
    "    \"\"\"\n",
    "      \n",
    "    parameters = {'s':s,'kp':kp, 'km':km, 'rho':rho}\n",
    "    simulation(production_decay_ode, parameters, initial_conditions, time)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "137ac0c4",
   "metadata": {},
   "source": [
    "#### Run the simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b574a39",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Define initial conditions and time intervals for the simulation\n",
    "initial_conditions = [0, 0] #[P, M]\n",
    "time = np.linspace(0, 100, 1001)\n",
    "\n",
    "interact(animate, \n",
    "         initial_conditions=fixed(initial_conditions), \n",
    "         time=fixed(time), \n",
    "         s = (1,10,0.01), #  Transcription rate units: concentration [P] per min    \n",
    "         kp = (1/120,1/10,1/1000), # Pre-mRNA degradation rate units: 1/min (range: 10 min to 120 min)\n",
    "         km = (1/600,1/10,1/1000),  # mRNA degradation rate units: 1/min (range: 10 min to 600 min)\n",
    "         rho = (1/30,1,0.05)); #splicing rate units: 1/min (range: 1 min to 30 min)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13f3fc7d",
   "metadata": {},
   "source": [
    "## Exercise 2: RNA-seq analysis and modelling of gene expression response to temperature up- and down-shifts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1c7c93e-1fdf-46fb-aba0-cadb6fdd2d00",
   "metadata": {},
   "source": [
    "In [Gotic et al., 2015](http://genesdev.cshlp.org/content/30/17/2005), cells were harvested at various time points (0, 1, 3, 6, and 9 hours) after a temperature switch from 33°C to 38°C and vice versa, with duplicates subjected to total RNA-seq analysis. The resulting data were processed to assign reads to pre-mRNA and mRNA species and to obtain their respective levels. Here, we will analyze those data to study and model gene expression responses to temperature shifts.\n",
    "\n",
    "The functions provided below (run the cell) will assist you in the analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff41c2e1",
   "metadata": {},
   "source": [
    "#### Implementation of useful functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d19b9c57-bbe4-41c1-8f98-5305a902f724",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def subselect(data, feature, replicate, time, condition):\n",
    "\n",
    "    \"\"\"\n",
    "    Subselects samples from the full dataset based on the specified features, conditions, \n",
    "    replicates, and time points.\n",
    "\n",
    "    Parameters:\n",
    "    data: The full gene expression dataset.\n",
    "    feature: The features to filter by (e.g., ['intron', 'exon']).\n",
    "    replicate: The replicates to include.\n",
    "    time: The time points to include.\n",
    "    condition: The temperature shift conditions to include (e.g, ['33to38', '38to33']).\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: The filtered dataset.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Create boolean masks for each filtering criterion\n",
    "\n",
    "    feature_mask  = data.columns.get_level_values('feature').isin(feature)\n",
    "    condition_mask  = data.columns.get_level_values('condition').isin(condition)\n",
    "    replicate_mask  = data.columns.get_level_values('replicate').isin(replicate)\n",
    "    time_mask = data.columns.get_level_values('time').isin(time)\n",
    "\n",
    "    # Use the masks to filter the data\n",
    "\n",
    "    filtered_data = data.iloc[:, feature_mask & condition_mask & replicate_mask & time_mask]\n",
    "    \n",
    "    return filtered_data\n",
    "\n",
    "def run_PCA(data, n_components=5, color_by='replicate', symbol_by='time', scale_data=True, log_transform=True):\n",
    "    \n",
    "    \"\"\"\n",
    "    Runs PCA on the provided data and generates a scatter plot of pairs of principal components.\n",
    "    It also returns a DataFrame containing the PCA loadings.\n",
    "\n",
    "    Parameters:\n",
    "    data: The gene expression data to analyze.\n",
    "    n_components: The number of principal components to calculate.\n",
    "    color_by: The metadata category to color the data points by.\n",
    "    symbol_by: The metadata category to symbolize the data points by.\n",
    "    scale_data: Whether to scale the data before PCA.\n",
    "    log_transform: Whether to apply log transformation to the data.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: A DataFrame containing the PCA loadings.\n",
    "    \"\"\"\n",
    "\n",
    "    data_tmp = data.copy()\n",
    "    \n",
    "    # Scale and/or log-transform the data if specified\n",
    "    if scale_data:\n",
    "        data = scale(data)\n",
    "    else:\n",
    "        data = data\n",
    "        \n",
    "    if log_transform:\n",
    "        data = np.log2(data + 1)\n",
    "        \n",
    "    # Fit the PCA model\n",
    "    pca = PCA(n_components=n_components)\n",
    "    pca.fit(data.T)\n",
    "\n",
    "    # Transform the data\n",
    "    pca_transformed = pca.transform(data.T)\n",
    "\n",
    "    # Create a DataFrame for the PCA results\n",
    "    labels = [f'PC{i+1}' for i in range(n_components)]\n",
    "    pca_df = pd.DataFrame(pca_transformed, columns=labels)\n",
    "    pca_df[color_by] = data_tmp.columns.get_level_values(color_by)\n",
    "    pca_df[symbol_by] = data_tmp.columns.get_level_values(symbol_by)\n",
    "\n",
    "    # Plotting each pair of consecutive principal components\n",
    "    fig, axes = plt.subplots(1, n_components -1, figsize=(30, 5))\n",
    "    for i in range(n_components-1):\n",
    "        sns.scatterplot(ax=axes[i], data=pca_df, x=f'PC{i+1}', y=f'PC{i+2}', \n",
    "        hue=color_by, style=symbol_by, s=200, palette='Set1')\n",
    "\n",
    "        axes[i].set_xlabel(f'PC{i+1} ({pca.explained_variance_ratio_[i]*100:.2f}%)')\n",
    "        axes[i].set_ylabel(f'PC{i+2} ({pca.explained_variance_ratio_[i+1]*100:.2f}%)')\n",
    "        # remove legend if it's not the first plot\n",
    "        if i != 0:\n",
    "            axes[i].get_legend().remove()\n",
    "    \n",
    "    fig.suptitle(f'PCA of Gene Expression Data (color={color_by}, symbol={symbol_by})', fontsize=16)\n",
    "\n",
    "    # Return the PCA loadings\n",
    "    df_pca = pd.DataFrame(pca.components_, columns=data_tmp.index, index=labels)\n",
    "\n",
    "    return df_pca\n",
    "\n",
    "\n",
    "## Plot Gene profile across all conditions\n",
    "def plot_gene(data, gene , xx_33_38=None, xx_38_33=None, t_2=None):\n",
    "\n",
    "    \"\"\"\n",
    "    Plots the gene expression profile across different conditions and time points.\n",
    "    \n",
    "    Parameters:\n",
    "    data: The gene expression data.\n",
    "    gene: The gene to plot.\n",
    "    xx_33_38: The simulated solution for the 33to38 condition.\n",
    "    xx_38_33: The simulated solution for the 38to33 condition.\n",
    "\n",
    "    Returns:\n",
    "    None\n",
    "    \"\"\"\n",
    "\n",
    "    #subselect\n",
    "    time= ['0', '1' ,'3' ,'6', '9']\n",
    "    replicate = ['2', '13', '11', '5']\n",
    "    dat_1= subselect(data, ['intron'], replicate, time, ['33to38'])\n",
    "    dat_2= subselect(data, ['exon'], replicate, time, ['33to38'])\n",
    "    dat_3= subselect(data, ['intron'], replicate, time, ['38to33'])\n",
    "    dat_4= subselect(data, ['exon'], replicate, time, ['38to33'])\n",
    "\n",
    "    #plot\n",
    "    t=np.array([0, 0, 1, 1, 3, 3, 6, 6, 9 , 9])\n",
    "    fig, axs = plt.subplots(2, 2, figsize=(8,8), sharex=True, sharey=True)\n",
    "    #fig.subplots_adjust(wspace=0.25)\n",
    "    axs[0, 0].scatter(t, dat_1[dat_1.index.str.endswith(gene)].values.T, c = [0,2]*5, vmin=0, vmax=3, label='intron 33to38') #color is replicate\n",
    "    axs[0, 0].set_title(\"intron 33to38\")\n",
    "    \n",
    "    axs[1, 0].scatter(t, dat_2[dat_2.index.str.endswith(gene)].values.T, c = [0,2]*5, vmin=0, vmax=3,  label='exon 33to38') #color is replicate\n",
    "    axs[1, 0].set_title(\"exons 33to38\")\n",
    "    \n",
    "    axs[0, 1].scatter(t, dat_3[dat_3.index.str.endswith(gene)].values.T, c = [1,3]*5, vmin=0, vmax=3, label='intron 38to33') #color is replicate\n",
    "    axs[0, 1].set_title(\"intron 38to33\")\n",
    "\n",
    "    axs[1, 1].scatter(t, dat_4[dat_4.index.str.endswith(gene)].values.T, c = [1,3]*5, vmin=0, vmax=3,  label='exon 38to33') #color is replicate\n",
    "    axs[1, 1].set_title(\"exon 38to33\")\n",
    "\n",
    "    if xx_33_38 is not None:\n",
    "        axs[0,0].plot(t_2, xx_33_38[:,0],label = 'P(t)', color=\"tab:orange\")\n",
    "        axs[1,0].plot(t_2, xx_33_38[:,1], color=\"tab:orange\")\n",
    "        axs[0,1].plot(t_2, xx_38_33[:,0], color=\"tab:orange\")\n",
    "        axs[1,1].plot(t_2, xx_38_33[:,1], color=\"tab:orange\")\n",
    "    \n",
    "        \n",
    "    for ax in axs.flat:\n",
    "        ax.set(xlabel='Time [h]', ylabel='RPKM')\n",
    "\n",
    "    fig.suptitle(f'Gene Expression Profile for {gene}', fontsize=16)\n",
    "    plt.show()\n",
    "\n",
    "def plot_DE(dat_all, condition_x1, condition_x2, condition_y1, condition_y2, time_x1, time_x2, time_y1, time_y2, feature_x1, feature_x2, feature_y1, feature_y2, FC=2, xlab=\"\", ylab=\"\", FC_type=\"delta\"):\n",
    "    \n",
    "    \"\"\"\n",
    "    Plots differentially expressed genes based on log fold changes between specified conditions, times, and features.\n",
    "\n",
    "    Parameters:\n",
    "    dat_all: The complete dataset with gene expression values.\n",
    "    condition_x1, condition_x2, condition_y1, condition_y2: Lists of conditions to compare.\n",
    "    time_x1, time_x2, time_y1, time_y2: Lists of time points to compare.\n",
    "    feature_x1, feature_x2, feature_y1, feature_y2: Lists of features to compare.\n",
    "    FC: Absolute fold change threshold for considering a gene differentially expressed in at least one of the two axes.\n",
    "    xlab: Label for the x-axis representing the log fold change between condition_x2 and condition_x1.\n",
    "    ylab: Label for the y-axis representing the log fold change between condition_y2 and condition_y1.\n",
    "\n",
    "    Returns:\n",
    "    Array: Boolean array indicating differentially expressed genes that meet the fold change threshold.\n",
    "    \"\"\"\n",
    "     \n",
    "    replicate = ['2', '13', '11', '5']\n",
    "\n",
    "    dat_x1=subselect(dat_all, feature_x1, replicate, time_x1, condition_x1).mean(axis=1)\n",
    "    dat_x2=subselect(dat_all, feature_x2, replicate, time_x2, condition_x2).mean(axis=1)\n",
    "\n",
    "    dat_y1=subselect(dat_all, feature_y1, replicate, time_y1, condition_y1).mean(axis=1)\n",
    "    dat_y2=subselect(dat_all, feature_y2, replicate, time_y2, condition_y2).mean(axis=1)\n",
    "\n",
    "    x=np.log2(1 + dat_x2)- np.log2(1 + dat_x1)\n",
    "    y=np.log2(1 + dat_y2) - np.log2(1 + dat_y1)\n",
    "\n",
    "    if FC_type == \"delta\":\n",
    "        pos =((x > FC) & (y > FC)) | ((x < -FC) & (y < -FC))\n",
    "\n",
    "    if FC_type == \"delta_delta\":\n",
    "        pos = abs(x-y) > FC\n",
    "\n",
    "    x_2 = x[pos]\n",
    "    y_2 = y[pos]\n",
    "    fig, ax = plt.subplots(figsize=(10,10))\n",
    "\n",
    "    ax.plot(x,y, 'o', color = 'lightblue', markersize=2)\n",
    "    ax.axhline(y=0, color='b')\n",
    "    ax.axvline(x=0, color='b')\n",
    "    ax.set_xlim(-8,8)\n",
    "    ax.set_ylim(-8,8)\n",
    "    ax.set_xlabel(xlab)\n",
    "    ax.set_ylabel(ylab)\n",
    "    texts = [ax.text(x_2.iloc[k], y_2.iloc[k], v.split(\"|\")[1]) for k, v in enumerate(dat_all.index[pos])]\n",
    "    adjust_text(texts)\n",
    "    \n",
    "    return(pos)    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75b39029-fddc-4e12-900e-332a6ca4851c",
   "metadata": {},
   "source": [
    "###  Question 1\n",
    "\n",
    "**PCA** (Read carefully)\n",
    "\n",
    "*\"Principal Component Analysis (PCA) is a dimensionality reduction technique that simplifies complex datasets by transforming them into a set of orthogonal axes, or principal components (PCs), that summarize the underlying patterns of variation. It works by calculating the eigenvectors and eigenvalues from the covariance matrix of the data, which then define the new axes. The first principal component captures the largest amount of variance in the data, with each subsequent component—perpendicular to the last—capturing progressively less variance. This orthogonal nature of PCA ensures that each component adds distinct information. This means that all the PCs are uncorrelated to each other. When applied to gene expression data, PCA reduces the numerous gene variables to a few composite indicators without losing critical information. After transformation, each principal component represents a combination of genes. The loadings of these components then tell us about the contribution of each gene to the component. Genes with high absolute loadings have a greater role in defining the structure revealed by that component, allowing us to identify which genes are most responsible for differentiating the samples on the basis of the observed variance.\"*\n",
    "  \n",
    "**GO enrichment analysis (GEA)**\n",
    "\n",
    "*\"Gene Ontology (GO) analysis is a common bioinformatics technique used to gain insight into the biological significance behind large lists of genes. When we perform experiments that result in the identification of many genes, such as with RNA-seq, we often want to understand the underlying biology of these genes. However, looking at each gene individually can be overwhelming due to the sheer number. This is where GO analysis comes in handy. GO analysis categorizes genes into groups known as \"GO terms\" describing different aspects of molecular biology or pathways. Using GO analysis, we can identify which biological processes are enriched in our gene list—meaning that these categories contain more genes from our list than we would expect by chance. This can suggest that the genes are collectively involved in certain biological functions or pathways, which may be crucial to the condition or treatment we're studying.\"*\n",
    "  \n",
    "Using the provided code below (PCA and enrichR cells), explore the RNA-Seq data:\n",
    "\n",
    "1. Perform principal component analysis (PCA) selecting different temperature conditions, time points, replicates and features (i.e. intron or exon). For each scenario described below, explore the principal components (PCs) and comment:\n",
    "  \n",
    "    a. With all samples included, determine the primary sources of variance. Is there an outlier? Do replicates cluster as expected?  \n",
    "    b. Focus on exonic counts and identify which principal component correlates with time.  \n",
    "    c. In the PCA representation, describe the transcriptome response over time. Are there notable differences between the two conditions? \n",
    "\n",
    "2. Extract gene PCA loadings (positive or/and negative) from relevant sample groups. \n",
    "\n",
    "    a. Identify genes functions that are temperature-regulated.  \n",
    "    b. Compare cold-induced versus heat-induced gene functions. What differences do you find?    \n",
    "    c. Advanced (Optional): Identify and discuss the main sources of variance between replicates.\n",
    "\n",
    "*Modify only the code marked as 'PCA cell' and 'enrichR' below*."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d96361a8-96c3-4a88-87ff-adfdb3207780",
   "metadata": {},
   "source": [
    "### Answer 1\n",
    "\n",
    "1. *Type your answers here*   \n",
    "\n",
    "a.   \n",
    "b.  \n",
    "c.  \n",
    "\n",
    "2. *Type your answers here*   \n",
    "\n",
    "a.   \n",
    "b.  \n",
    "c.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86df6bf4",
   "metadata": {},
   "source": [
    "#### Load the RNA-seq data (normalized counts) and format it as a panda data.frame\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "577eb2a7-9ada-49df-b4c4-9fb9162ee277",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load 33to38 data\n",
    "dat_33_38 = pd.read_csv(\"./GoticData/GSE85553_33-38_exon_intron_RPKM.txt\",sep='\\t')\n",
    "dat_33_38=dat_33_38.add_prefix(\"33to38.\")\n",
    "\n",
    "#Load 33to38 data\n",
    "dat_38_33 = pd.read_csv(\"./GoticData/GSE85553_38-33_exon_intron_RPKM.txt\",sep='\\t')\n",
    "dat_38_33=dat_38_33.add_prefix(\"38to33.\")\n",
    "\n",
    "#concatenate\n",
    "dat_all = pd.concat([dat_33_38, dat_38_33.iloc[:, :20]],axis=1)\n",
    "dat_all.columns=dat_all.columns.str.replace('\\.t','.', regex=True)\n",
    "dat_all.replace([np.inf, -np.inf], 0, inplace=True)\n",
    "\n",
    "#create column names\n",
    "dat_all.columns=dat_all.columns.str.split('.', expand = True)\n",
    "dat_all.columns.names = ['condition','time','replicate','feature']\n",
    "\n",
    "# have a look at the data\n",
    "with pd.option_context('display.max_columns', None):\n",
    "    print(dat_all.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b64fd718",
   "metadata": {},
   "source": [
    "#### Keep expressed genes (i.e. genes with mean expression > 1 and > 10 RPKM in intron and exon quantification, respectively)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87113bef-dd6d-4add-a465-7ac782d4373e",
   "metadata": {},
   "outputs": [],
   "source": [
    "time= ['0', '1' ,'3' ,'6', '9']\n",
    "replicate = ['2', '13', '11', '5']\n",
    "conditions = ['33to38','38to33']\n",
    "\n",
    "dat_intron= subselect(dat_all, ['intron'], replicate, time, conditions)\n",
    "dat_exon= subselect(dat_all, ['exon'], replicate, time, conditions)\n",
    "dat_all = dat_all.loc[(dat_intron.mean(axis=1) > 1 ) & (dat_exon.mean(axis=1) > 10), :]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21dac8cb",
   "metadata": {},
   "source": [
    "### PCA cell\n",
    "#### Perform PCA selecting different temperature conditions, timepoints, replicate and features (intron vs. exon). Change those parameters to select different samples for the PCA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03cf66ae-b8ad-4b7c-8949-529110ac0a61",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#Choose the features, conditions, replicates, and time points to include in the analysis for the different questions.\n",
    "condition = ['38to33','33to38'] \n",
    "time= ['0', '1' ,'3' ,'6', '9']\n",
    "replicate = ['2', '13', '11', '5'] # replicates 13 and 2 are from the 33to38 experiment, 5 and 11 from the 38to33.\n",
    "feature = ['intron']\n",
    "\n",
    "#Subselect the samples \n",
    "dat_sub = subselect(dat_all, feature, replicate, time, condition)\n",
    "\n",
    "#Perform the PCA, plot, and return the PCA loadings. Adjust the parameters color_by and symbol_by to answer to the different question.\n",
    "PC_loadings_time_replicate = run_PCA(dat_sub, n_components=5, color_by='time', symbol_by='replicate', scale_data=True, log_transform=True) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a010f23d-8b24-4cc8-b868-cde1f27661d5",
   "metadata": {},
   "source": [
    "### Question 2\n",
    "In this exercise, we will explore the RNA-seq data to find genes whose expression levels change in response to temperature shifts (i.e. *Cirbp* is known to be cold-inducible). We will use an approach to directly identify genes that are differentially expressed (DE) when cells are moved from one temperature to another. Modify only the *DE cell* below.\n",
    "\n",
    "1.1. Use the function *plot_DE* to visualize genes with a change in expression between 33°C and 38°C. Look for genes with patterns similar to or opposite that of *Cirbp* — that is, those that are upregulated (cold-inducible) or downregulated (heat-inducible) when temperature decreases. \n",
    "\n",
    "1.2. Using the *plot_gene* function, show the temporal dynamic of genes identified as having temperature-dependent regulation of gene expression.\n",
    "\n",
    "\n",
    "2.1. Use the function *plot_DE* to highlight genes that are regulated after transcription (i.e. post-transcriptionally) by temperature changes. These genes may show differential expression at the mRNA level that are not seen at the pre-mRNA level. \n",
    "\n",
    "2.2. Using the *plot_gene* function, show the temporal dynamic of some of those genes. What can you say in terms of transcriptional versus post-transcriptional regulation?  \n",
    "\n",
    "\n",
    "*Hint: Be aware that the FC_type argument of the plot_DE function allows you to either:*\n",
    "- *Highlight genes with 'extreme' fold changes (FC_type = 'delta')*\n",
    "- *Highlight genes that out of the diagonal (FC_type = 'delta_delta')*\n",
    "\n",
    "3.1. Looking at the temporal dynamic of Cirbp during temperature shifts at the intronic and exonic level, what can you say about its regulation?  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "798ece84",
   "metadata": {},
   "source": [
    "### Answer 2  \n",
    "1.1. *Type your answer here*  \n",
    "\n",
    "1.2. *Type your answer here*  \n",
    "\n",
    "2.1. *Type your answer here*  \n",
    "\n",
    "2.2. *Type your answer here*  \n",
    "\n",
    "3.1. *Type your answer here*  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3ba7095-6e1d-4917-9145-191042b49814",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Running example, modify it to answer the different questions:\n",
    "DE_genes = plot_DE(dat_all, condition_x1=['33to38'], condition_x2=['33to38'], condition_y1=['38to33'], condition_y2=['38to33'], # modify the condition parameters ['33to38'], ['38to33']\n",
    "                time_x1=['0'], time_x2=['1'], time_y1=['1'], time_y2=['0'], # modify the time parameters ['0'], ['1'], ['3'], ['6'], ['9']\n",
    "                feature_x1=['exon'], feature_x2=['exon'], feature_y1=['exon'], feature_y2=['exon'], # modify the feature parameters ['intron'], ['exon']\n",
    "                FC=1, # adapt the FC threshold for readibility\n",
    "                xlab='38C(t=1) - 33C(t=0) (33to38 condition, Log2FC)',\n",
    "                ylab='38C(t=0) - 33C(t=1) (38to33 condition, Log2FC)',\n",
    "                FC_type=\"delta\") # this allows to select the type of fold change to use for the plot (\"delta\" or \"delta_delta\", extremities or out-of-diagonal)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c13f9579-b4c4-4860-ab16-cbafc6bf66e8",
   "metadata": {},
   "source": [
    "### Question 3\n",
    "\n",
    "We will now simulate the production decay model from Question 1 again, but with a temperature shift (i.e. temperature-dependent parameters).\n",
    "\n",
    "1. Using the widget provided below, simulate the ODE with distinct parameters for the two temperatures. Comment on the outcome.\n",
    "\n",
    "2. Use the widget to determine the parameters ($\\rho_{33}$, $\\rho_{38}$, $k_{m,33}$, $k_{m,38}$) that best fit *Cirbp* gene expression profile. Comment on your result.  \n",
    "\n",
    "3. *Optional*: replace *Cirbp* with other genes identified previously as regulated post-transcriptionally. You might want to change the parameters and limits of the y-scale. Comment on your results. \n",
    "\n",
    "\n",
    "*Note: For simplification, compared to the model in the paper, we have used a model without the $\\alpha$ parameter, which is still adequate for fitting most gene expression profiles. However, a limitation of this simplified model is that the obtained parameters may not be entirely realistic, which is the argument the authors used to favor the model with the $\\alpha$ parameter.*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40318be6",
   "metadata": {},
   "source": [
    "#### Implementation of the functions for the simulation with a temperature shift"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1171ac59",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_ss_param(s, kp, km, rho):\n",
    "    \"\"\"\n",
    "    Compute the steady-state concentrations of pre-mRNA and mRNA (P_ss and M_ss) given the parameters of the model.\n",
    "\n",
    "    \"\"\"\n",
    "    P_ss = s / (kp + rho)\n",
    "    M_ss = (rho / km) * (s / kp + rho)\n",
    "    return [P_ss, M_ss]  \n",
    "\n",
    "def simulation(ode_function, parameters, initial_conditions = None, time = None):\n",
    "\n",
    "    \"\"\"\n",
    "    Simulate the ODEs for the production and decay of pre-mRNA and mRNA.\n",
    "    parameters: parameters for the ODEs\n",
    "    initial_conditions: initial conditions for the ODEs\n",
    "    time: time intervals for the simulation\n",
    "    \"\"\"\n",
    "\n",
    "    s=parameters['s']\n",
    "    kp=parameters['kp']\n",
    "    km=parameters['km']\n",
    "    rho=parameters['rho']\n",
    "\n",
    "    xx=odeint(ode_function, initial_conditions, time, args=(s, kp, km, rho))\n",
    "    \n",
    "    return(xx)\n",
    "\n",
    "def animate(time, s_33=5560, s_38=5560, km_33=0.1, km_38=0.1, kp= 20.794, rho_33=0.05, rho_38=0.05, gene='Cirbp'):\n",
    "    \n",
    "    \"\"\"\n",
    "    Runs the simulation with interactive widgets for parameter exploration and using the steady-state of each temperature condition as initial condition to simulate the switch.\n",
    "    \"\"\"\n",
    "\n",
    "    #33 to 38\n",
    "    initial_conditions = compute_ss_param(s_33, kp, km_33, rho_33)\n",
    "    parameters = {'s':s_38,'km':km_38, 'kp':kp, 'rho':rho_38}\n",
    "    xx_33_38 = simulation(production_decay_ode, parameters, initial_conditions, time)\n",
    "    \n",
    "    #38 to 33\n",
    "    initial_conditions = compute_ss_param(s_38, kp, km_38, rho_38)\n",
    "\n",
    "    parameters = {'s':s_33,'km':km_33, 'kp':kp, 'rho':rho_33}\n",
    "    xx_38_33 = simulation(production_decay_ode, parameters, initial_conditions, time)\n",
    "    \n",
    "   \n",
    "    plot_gene(dat_all, gene, xx_33_38, xx_38_33, time)    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "deb2f6bd",
   "metadata": {},
   "source": [
    "#### Run the simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed619ea6-d668-4101-a549-d5d8e3f9deb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "t = np.linspace(0,9,9*3600)\n",
    "interact(animate, time=fixed(t), gene = 'Cirbp',\n",
    "         s_33 = (500,10000,1), # [RPKM/hr] at 33C\n",
    "         s_38 = (500,10000,1), # [RPKM/hr] at 38C\n",
    "         km_33 = (0.05,2,0.05), # [hr^-1], you have to find the correct km_33 \n",
    "         km_38 = (0.05,2,0.05), # [hr^-1], you have to find the correct km_38 \n",
    "         kp = (0.05,24,0.1),  # [hr^-1], around 2min for Cirbp\n",
    "         rho_33 = (0.05,5,0.1), # [hr^-1],  you have to find the correct rho_33 \n",
    "         rho_38 = (0.05,5,0.1)); # [hr^-1], you have to find the correct rho_38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0f44bbb-6e65-4923-be31-53222900e554",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
